### PR TITLE
Uploading apps (Fix #10)

### LINKED ISSUE #10: Uploading An App Release
PUT /v1/apps/9.0.0 with the JSON contents of {"download": "https://domain.com/path/to/tar.gz"} should download an app zip, parse the info.xml and update/create a new app release.

### PR DESCRIPTION
**Pull request by BernhardPosselt**
[Rendered](https://github.com/nextcloud/appstore/blob/upload-app/docs/api.md)

Todo List:
- [x] Download the release archive and make sure its not bigger than 50mb
- [x] Extract tar.gz and validate the structure
- [x] Validate XML
- [x] Parse XML into a dict
- [x] Validate the app release version by comparing it to the URL
- [x] Generate a sha256 over the release and delete the downloaded tar file
- [x] Test that the db transaction rolls back in case of errors
- [x] Map the parsed dict onto models
- [x] validate and save the models
- [x] Put the pieces together

### CODE REVIEW

File: nextcloudappstore/core/models.py [hunk 1]
```diff
@@ -39,6 +39,10 @@ class App(TranslatableModel):
                                             verbose_name=_('Co-Maintainers'),
                                             related_name='co_maintained_apps',
                                             blank=True)
+    recommendations = models.ManyToManyField('auth.User',
```
Comment (adsworth) [inline nextcloudappstore/core/models.py]: We should user `settings.AUTH_USER_MODEL` for ForeigKeys and ManyToManyfields.
See https://docs.djangoproject.com/en/1.9/topics/auth/customizing/#referencing-the-user-model

Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: Right, you're in the wrong PR though, app upload is not finished yet ;)

Can you create a new ticket?

**Comment by BernhardPosselt** [comment]
@adsworth @LukasReschke read to review \o/

**Comment by BernhardPosselt** [comment]
Can be tested with curl:

Create Minimal News release:

    curl -X POST -u "admin:admin" http://localhost:8000/api/v1/apps/releases -H "Content-Type: application/json" -d '{"download":"https://github.com/nextcloud/appstore/raw/c8d17ef91ffd04fd5792f5832ef7d77d9de88d1d/nextcloudappstore/core/api/v1/tests/data/archives/minimal.tar.gz"}'

Check the output:
    
    curl http://localhost:8000/api/v1/platform/9.0.0/apps.json | python -m json.tool

Update it:

   curl -X POST -u "admin:admin" http://localhost:8000/api/v1/apps/releases -H "Content-Type: application/json" -d '{"download":"https://github.com/nextcloud/appstore/raw/c8d17ef91ffd04fd5792f5832ef7d77d9de88d1d/nextcloudappstore/core/api/v1/tests/data/archives/full.tar.gz"}'

File: nextcloudappstore/core/api/v1/release/parser.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    pass
+
+
+class GunZipAppMetadataExtractor:
+    def __init__(self, config: ReleaseConfig) -> None:
+        """
+        :argument config the config
+        """
+        self.config = config
+        self.app_folder_regex = re.compile(r'^[a-z]+[a-z_]*$')
```
Comment (adsworth) [inline nextcloudappstore/core/api/v1/release/parser.py]: any particular reason we don't support numbers in app names?

File: nextcloudappstore/core/api/v1/release/parser.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    type = element.get('type')
+    key = element.tag.replace('-', '_')
+    if type == 'int':
+        return {key: int(element.text)}
+    elif type == 'list':
+        return {key: list(map(element_to_dict, element.iterchildren()))}
+    elif len(list(element)) > 0:
+        contents = {}
+        for child in element.iterchildren():
+            contents.update(element_to_dict(child))
```
Comment (adsworth) [inline nextcloudappstore/core/api/v1/release/parser.py]: Would it make sense to add some sort of max recursion depth?

**Comment by adsworth** [comment]
Looks good to me. Though I didn't check the xslt and xsd.

File: nextcloudappstore/core/api/v1/release/parser.py (hunk 2 above)
Comment (BernhardPosselt) [inline nextcloudappstore/core/api/v1/release/parser.py]: ownCloud can only deal with these characters for apps. PHP autoloader suckage :)

File: nextcloudappstore/core/api/v1/release/parser.py (hunk 3 above)
Comment (BernhardPosselt) [inline nextcloudappstore/core/api/v1/release/parser.py]: The Schema should not allow that much recursion. I think we are below 10

**Comment by BernhardPosselt** [comment]
Yep, thats arcane knowledge ;)

**Pull request merged**